File: 4. Shell Sort - GFG/4.-shell-sort.py
#User function Template for python3

def shellSort(arr, n):
    # code here
    gap = n//2
    while gap>0:
        for i in range(gap,n):
            x = arr[i]
            j = i
            while j>= gap and x<arr[j-gap]:
                arr[j] = arr[j-gap]
                j = j - gap
            arr[j] = x
        gap = gap//2
    return arr    

#{ 
#  Driver Code Starts
#Initial Template for Python 3


if __name__ == "__main__":
    t = int(input())
    while(t>0):
        n = int(input())
        arr = [int(x) for x in input().strip().split()]
        shellSort(arr, n)
        for ele in arr:
            print(ele, end=" ")
        print()
        t=t-1

# } Driver Code Ends

File: Sorting_DSA_Basics/4. Shell Sort - GFG/4.-shell-sort.cpp
// { Driver Code Starts
//Initial Template for C++

#include  <bits/stdc++.h> 
using namespace std; 


 // } Driver Code Ends
//User function Template for C++

void shellSort(int arr[], int n) 
{
    // code here 
    for(int gap = n/2; gap >=1; gap/=2){
        for(int j = gap; j < n; j++){
            for(int i = j - gap; i>=0; i-=gap){
                if(arr[i+gap] > arr[i]) break;
                else swap(arr[i],arr[i+gap]);
            }
        }
    }
}

// { Driver Code Starts.

int main() 
{ 
    int t;
    cin>>t;
    while(t--)
    {
        int n;
        cin>>n;
        int arr[n];
        
        for(int i=0;i<n;i++)
            cin>>arr[i];
  
        shellSort(arr, n); 
        
        for(int i=0;i<n;i++)
            cout<<arr[i]<<" ";
        
        cout<<endl;
    }
    return 0; 
}   // } Driver Code Ends